### ISSUE TITLE
Proposal: Remove Locale and java.time from Scala Native core

### ISSUE THREAD

**Issue by ekrich** [scala-native/scala-native#1903]
In order to pave the way for third party libraries to support `Locale` and `java.time` we should remove these APIs from Scala Native.

This will more closely align Scala Native to Scala.js and make it possible to have cross compiling libraries to support these APIs.

Here is a list of issues that can be addressed via this meta issue.

[ ] #595
[ ] #999 
[ ] #1038 
[ ] #1443 
[ ] Agree upon the API approach to use for Scala Native and Scala.js for Java where part of the implementation is in Scala Native and Scala.js and the other methods are in the third party library. See https://github.com/cquiroz/scala-java-locales/pull/199#issuecomment-687542408 This will need to be approved by the Scala Center to assure a uniform approach.

I think it is pretty well agreed upon to pursue this path. We have the following 2 libraries that could support this effort.

https://github.com/cquiroz/scala-java-locales/
https://github.com/cquiroz/scala-java-time/

**Comment by er1c** [comment]
One of the "biggest" (is it that big?) problems with this proposal is String.toLowerCase(Locale) won't be  supported, so it would need to be mixed-in as some sort of `.toLowerCaseCompat(Locale)` method name if I understand a comment made by @sjrd here: https://github.com/cquiroz/scala-java-locales/pull/199#issuecomment-687542408 I'm not sure if that's the same thing with scala-native

**Comment by LeeTibbert** [comment]
In support of the above, I have most of scala-java-time working with SN 0.4.0-SNAPSHOT, circa Sept2020.
A variant of the sjt demo project lightly modified for SN executes as expected.

I also have large parts of scala-java-locales working.

The piece blocking my submitting PRs to sjt etc is that I do not yet have ScalaTest working with 
current 0.4.0-SNAPSHOT.  I have the Logger & Config changes working, but am spending
weeks on getting Reflection working.  Yeah, I know, it ought to be easy, it just make it like
Scala.js.   I have a query in the ScalaTest gitter out for one of the two issues which are
hosing me (mumble$ReflectivelyInstantiate not getting into the artifact jar).

**Comment by LeeTibbert** [comment]
@er1c Could you create a checkbox so that we can number the beasts and begin to get an estimate of "size of effort" and what needs further __focused__ design?   I must confess that some of these issues are too complex for me to properly appreciate on the fly.

**Comment by er1c** [comment]
@LeeTibbert 

> @er1c Could you create a checkbox so that we can number the beasts and begin to get an estimate of "size of effort" and what needs further **focused** design? I must confess that some of these issues are too complex for me to properly appreciate on the fly.

me or @ekrich ? :)

**Comment by er1c** [comment]
FYI: https://github.com/scala-js/scala-js/pull/4214

**Comment by sjrd** [comment]
So, in the past few days, I have been thinking about a solution to this issue. I've used Scala.js as a playground because I could easily *add* things there to see if things could work. It turns out that they can.

For background, in Scala.js we have a stricter policy about what goes into the core. Our policy, which is (almost) a hard rule, is: if it links, it is correct. Based on that policy, we had so far:

* Refused to implement anything based on `java.util.Locale` in the core, because that would require depending on CLDR databases, and it's a huge thing to maintain, which we cannot do.
* Therefore, refused to implement `java.lang.String.to{Lower,Upper}Case(Locale)`
* as well as `Locale`-specific features of `java.util.Formatter`.

In addition, and to be able to correctly support methods relying on the "default" Locale (such as `java.lang.String.to{Lower,Upper}Case()`, or constructors of `j.u.Formatter` without an explicit Locale, we have decreed that the default locale of Scala.js is always en-US, and that is not settable. Under that decree, we could hard-code en-US behaviors in our library, and live up to our policy while allowing many use cases to work.

Meanwhile, for classes such as `j.u.Locale` and `java.text.*`, we encouraged developers to create separate libraries to implement them. This resulted in the very successful scala-java-locales, mentioned above. This strategy is great for everything where the features are in separate classes: they can be implemented separately from the core, by people who have the necessary expertise and willingness to maintain locale-dependent features.

It did not work so great for individual *methods* that are locale-dependent and that live inside classes that are implemented in the core. The famous example being `String.to{Lower,Upper}Case(Locale)`:
* We can't test them in the core because we don't have `Locale`, so realistically we cannot implement them in the core.
* They cannot be implemented in a separate library because a separate library cannot add methods to classes that are part of the core.

In summary: they simply could not be implemented at all. And for years we have lived with the unfortunate situation that No, Scala.js users cannot get `String.to{Lower,Upper}Case(Locale)`.

Another example, less problematic, was the Locale-specific features of `java.util.Formatter`.

---

Meanwhile, in Scala Native, the policy was basically: do whatever, as long as it allows "things to work". This comes at a price: it is not guaranteed that if something links, it will be correct. But it comes with the ability to implement a few more things, notably the overloads of `toLowerCase`, `toUpperCase` and `Formatter` that take `Locale`s as parameters.

It also came with the price identified in this issue: even if a third-party library wants to correctly implement those features, it cannot, because it cannot supplant the half-baked (or should I say tenth-baked) implementations of classes like `java.text.*`.

So we want to remove them from the core. But if we do that, some code will break. And in particular most of the code of `j.u.Formatter` would not link anymore even in the Scala Native test suite itself. So we would have to rip out the Locale-dependent parts of `Formatter`, which means that we would break even more code in a way that cannot be recovered in a third-party.

This issue forced me to think hard about a way to resolve the situation, for good.

---

So, starting from Scala.js, which has a strict policy, I asked myself: how can I implement `String.toUpperCase(Locale)` *and test it* in the core, while not having `Locale`? Well, it turns out there was an easy solution:

* Create a "fake", dummy "third-party" implementation of `Locale` inside a separate project in the core repo
* Create a separate test suite that, in addition to relying on the javalib that we distribute, also depends on the above dummies
* In that test suite, test the Locale-specific methods

This strategy allows to:

* Keep the normal test suite free of the dummies. This is very important to make sure that everything *else* that we test keeps linking and working even if we don't have `Locale` on the classpath.
* Yet correctly test the Locale-specific methods.

I implemented the infrastructure for this strategy in https://github.com/scala-js/scala-js/pull/4211/commits/54648372c909167b7360285bf2ff25b91db42d96 for a trivial method, which I then used in https://github.com/scala-js/scala-js/pull/4214 to implement `String.to{Lower,Upper}Case(Locale)`.

For `java.util.Formatter`, it's a bit more tricky, because we need to "store" the `Locale` as a field of the `Formatter`, yet we can't actually do that because the test suite without the dummy Locale would cease to link (and so would any codebase that does not pull in scala-java-locales). The trick is to leverage the ability of the reachability analysis to reason about object-oriented hierarchies, and its ability to dead-code-eliminate subclasses that are never instantiated. This strategy led to my pending PR https://github.com/scala-js/scala-js/pull/4218.

---

Now that we have proved in Scala.js that we can do it, we simply have to adapt the strategy to Scala Native. For `toLowerCase` and `toUpperCase`, this will be trivial. For `Formatter`, it's still a bit more tricky because there is all the Time-dependent features, which was never implemented in Scala.js. But I'm pretty confident that the OO-based trick will work here as well.

---

So ... who wants to submit a PR? :sweat_smile:

**Comment by ekrich** [comment]
This is really great news.

To be fair, I don't think anyone realized early on the implications of adding `Locale` for `toUpperCase` and friends. At the time we were just trying to make things work and I don't think anyone thought about whether parts of the Java APIs should be part of the core or supplied by a third party. We also had a collusion of folks, me included, trying to get `scalafmt` running on Scala Native which added some of the `java.time` and friends creating more problems.

**Comment by lolgab** [comment]
I tried to update [scalanative-java-time](https://github.com/akka-js/scalanative-java-time) to 0.4.0-M2 and since there is already an implementation of some classes that throw `UnsupportedOperationException` in `javalib` it doesn't substitute them with the real implementations, making the library unusable.
Note that a minimal java-time is a dependency of [ZIO](https://zio.dev) which is a library nice to have on the next release.

**Comment by LeeTibbert** [comment]
@lolgab   A few months or so ago I had scala-java-time core working with the then SNAPSHOT of SN.
I think getting the scala-java-time tests to work took some PRs that are still in the queue.

I seem to remember that I had Zio working, or at least linking, with scala-java-time.

I seem to remember that @catap got ScalaTest (used by scala-java-time) to work with 0.4.0-SNAPSHOT.
I do not remember if he was also active with scala-java-time.

Now scalanative-java-time is fine, but I think the economics of scala-java-time, especially
its test suite, are better.   There is simply no way that a portable-scala-java-time or a scalanative-java-time
is going to catch up with the TZ support and test suite.

When I was using scalanative-java-time with ZIO, almost a year ago, I discovered the trick of
_prepending_  the scalanative-java-time dependency to get around the whole javalib time mess.
I _think_ but do not remember well that I was able to shadow things, such as Instant, which occur in
the SN javalib.  I can try to scratch up that few lines of code if you like.

These days I am only mildly adverse to ripping out all of javalib time.   I personally am more interested
in a minimally competent implementation, with a referral to one or more competent and assisted third party
mumble-java-time libraries for complicated stuff.

I have been pecking away at supplying the infrastructure routines in SN needed by scala-java-time, especially the testing.
To my thinking, SN needs at least one demonstrated, competent third-party before it can disable its javalib
time routines.

A few months or so, @ekrich and I did a syncopated walkthru of what it would take to remove the time routines
from SN. IIRC, we had slightly different approaches and opinions. It can be done, but I think there some places
which will take time to do without upsetting everything around it. I forget the routine, but there was something
which SN used itself.

I think I could fix up my File nanosecond timestamp PR to use its own private competent Instant.
